src/bitmap/bitmap_8k.rs

```diff
@@ -0,0 +1,314 @@
+        }
+    }
+}
+
+impl<B: Borrow<[u64; BITMAP_LENGTH]>> Iterator for BitmapIter<B> {
+    type Item = u16;
+
+    fn next(&mut self) -> Option<u16> {
+        loop {
+            if self.key == BITMAP_LENGTH {
+                return None;
+            } else if (unsafe { self.bits.borrow().get_unchecked(self.key) } & (1u64 << self.bit))
+                != 0
+            {
+                let result = Some((self.key * 64 + self.bit) as u16);
+                self.move_next();
+                return result;
+            } else {
+                self.move_next();
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        panic!("Should never be called (roaring::Iter caches the size_hint itself)")
+    }
+}
+
+#[inline]
+pub fn key(index: u16) -> usize {
+    index as usize / 64
+}
+
+#[inline]
+pub fn bit(index: u16) -> usize {
+    index as usize % 64
+}
+
+#[inline]
+fn op_bitmaps(bits1: &mut Bitmap8K, bits2: &Bitmap8K, op: fn(&mut u64, u64)) {
+    bits1.len = 0;
+    for (index1, &index2) in bits1.bits.iter_mut().zip(bits2.bits.iter()) {
+        op(index1, index2);
+        bits1.len += index1.count_ones() as u64;
+    }
+}
+
+impl BitOrAssign<&Self> for Bitmap8K {
+    fn bitor_assign(&mut self, rhs: &Self) {
+        op_bitmaps(self, rhs, BitOrAssign::bitor_assign);
+    }
+}
+
+impl BitOrAssign<&SortedU16Vec> for Bitmap8K {
+    fn bitor_assign(&mut self, rhs: &SortedU16Vec) {
+        for &index in rhs.iter() {
+            let (key, bit) = (key(index), bit(index));
+            let old_w = self.bits[key];
+            let new_w = old_w | 1 << bit;
+            self.len += (old_w ^ new_w) >> bit;
+            self.bits[key] = new_w;
+        }
+    }
+}
+
+impl BitAndAssign<&Self> for Bitmap8K {
+    fn bitand_assign(&mut self, rhs: &Self) {
+        op_bitmaps(self, rhs, BitAndAssign::bitand_assign);
+    }
+}
+
+impl SubAssign<&Self> for Bitmap8K {
+    fn sub_assign(&mut self, rhs: &Self) {
+        op_bitmaps(self, rhs, |l, r| *l &= !r);
+    }
+}
+
+impl SubAssign<&SortedU16Vec> for Bitmap8K {
+    fn sub_assign(&mut self, rhs: &SortedU16Vec) {
+        for &index in rhs.iter() {
+            let (key, bit) = (key(index), bit(index));
+            let old_w = self.bits[key];
+            let new_w = old_w & !(1 << bit);
+            self.len -= (old_w ^ new_w) >> bit;
+            self.bits[key] = new_w;
+        }
+    }
+}
+
+impl BitXorAssign<&Self> for Bitmap8K {
+    fn bitxor_assign(&mut self, rhs: &Self) {
+        op_bitmaps(self, rhs, BitXorAssign::bitxor_assign);
+    }
+}
+
+impl BitXorAssign<&SortedU16Vec> for Bitmap8K {
+    fn bitxor_assign(&mut self, rhs: &SortedU16Vec) {
+        let mut len = self.len as i64;
+        for &index in rhs.iter() {
+            let (key, bit) = (key(index), bit(index));
+            let old_w = self.bits[key];
+            let new_w = old_w ^ 1 << bit;
+            len += 1 - 2 * (((1 << bit) & old_w) >> bit) as i64; // +1 or -1
```

Comment:
Ho! So here is one way to remove some conditions, no more jumps for that!

Reply:
Credit to:
https://github.com/RoaringBitmap/CRoaring/blob/f046613e9b75f787b618e7b62197415ea7c2c879/src/bitset_util.c#L985